File: openworkflow-studio/webapp/src/main/webapp/src/workflow.ts
import { load } from "js-yaml";
import { ResponseError } from "@forwardmeasure/openworkflow-definition-management-client";

export const SAMPLE = `document:
  dsl: '1.0.3'
  namespace: forwardmeasure
  name: hello-studio
  version: '1.0.0'
do:
  - greet:
      set:
        message: Hello from OpenWorkflow Studio
`;

export function taskNames(source: string): string[] {
  try {
    const workflow = load(source) as { do?: Array<Record<string, unknown>> };
    return Array.isArray(workflow?.do)
      ? workflow.do.flatMap((task) => Object.keys(task))
      : [];
  } catch {
    return [];
  }
}

// The generated client's ResponseError only ever carries a generic message
// ("Response returned an error code") - the actual server-side explanation
// is in the response BODY, which nothing was reading. Two Problem shapes
// exist and both need reading, not just one: RFC 9457 Problems from
// openworkflow-definition-management-jaxrs's *ExceptionMapper classes carry
// the real text in "detail" (an exception message for known error types,
// and, temporarily, see DebugThrowableExceptionMapper - the full exception
// class/cause chain/stack trace for anything else); Quarkus's own built-in
// bean-validation failure response (quarkus-hibernate-validator rejecting a
// request before a resource method even runs, e.g. a missing required
// header) has no "detail" at all - the real text is in "violations"
// instead. Async because reading a Response body is a Promise; every call
// site already awaits inside a try/catch.
export async function diagnostic(error: unknown): Promise<string> {
  if (error instanceof ResponseError) {
    try {
      const problem = (await error.response.clone().json()) as {
        detail?: string;
        title?: string;
        violations?: Array<{ field?: string; message?: string }>;
      };
      if (problem.detail) {
        return problem.title
          ? `${problem.title}: ${problem.detail}`
          : problem.detail;
      }
      if (problem.violations?.length) {
        const summary = problem.violations
          .map(
            (violation) =>
              `${violation.field ?? "value"} ${violation.message ?? "is invalid"}`,
          )
          .join("; ");
        return problem.title ? `${problem.title}: ${summary}` : summary;
      }
      if (problem.title) return problem.title;
    } catch {
      // Response body wasn't JSON (or already consumed) - fall through to the generic message.
    }
  }
  if (error instanceof Error) return error.message;
  return String(error);
}

export function canPause(status: string): boolean {
  return status === "RUNNING" || status === "WAITING";
}

/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at https://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable
 * law or agreed to in writing, software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
 * for the specific language governing permissions and limitations under the License.
 */

File: openworkflow-definition-management/openworkflow-definition-management-application/src/main/java/com/forwardmeasure/openworkflow/definition/domain/entity/Workflow.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.forwardmeasure.openworkflow.definition.domain.entity;

import com.forwardmeasure.jpa.identity.entity.OwnedEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import java.io.Serial;
import java.util.Objects;
import java.util.regex.Pattern;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * The stable, named, owned workflow identity. Content and governance live on {@link
 * WorkflowDefinition}.
 */
@Entity
@Table(name = "workflow")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Workflow extends OwnedEntity<Long> {
  @Serial private static final long serialVersionUID = 1L;
  private static final Pattern NAME_PATTERN = Pattern.compile("[a-z][a-z0-9]*(?:-[a-z0-9]+)*");

  @Id
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "workflow_ids")
  @SequenceGenerator(name = "workflow_ids", sequenceName = "workflow_id_seq", allocationSize = 50)
  @Setter
  private Long id;

  @Column(name = "name", nullable = false, length = 128)
  private String name;

  @Column(name = "title", nullable = false, length = 512)
  private String title;

  @Column(name = "description", length = 4096)
  @Setter
  private String description;

  public Workflow(String name, String title, String description) {
    this.name = validateName(name);
    this.title = validateTitle(title);
    this.description = description;
  }

  public void setName(String name) {
    this.name = validateName(name);
  }

  public void setTitle(String title) {
    this.title = validateTitle(title);
  }

  private static String validateName(String value) {
    if (!NAME_PATTERN.matcher(required(value, "name")).matches()) {
      throw new IllegalArgumentException(
          "name must match " + NAME_PATTERN.pattern() + ": " + value);
    }
    return value;
  }

  private static String validateTitle(String value) {
    return required(value, "title");
  }

  private static String required(String value, String name) {
    Objects.requireNonNull(value, name);
    if (value.isBlank()) {
      throw new IllegalArgumentException(name + " must not be blank");
    }
    return value;
  }
}
